src/coingecko.ts

```diff
@@ -0,0 +1,220 @@
+    .readdirSync(BUILD_DIR)
+    .filter(
+      (file) => path.extname(file) === ".json" && file.startsWith(FILE_PREFIX)
+    );
+
+  tokenFiles.forEach((file) => {
+    const tokensJson = fs.readFileSync(path.join(BUILD_DIR, file));
+    const tokens = JSON.parse(tokensJson.toString()) as Token[];
+    combinedTokens.push(...tokens);
+  });
+
+  const filePath = path.join(LIST_DIR, OUTPUT_FILE);
+  await writeJson(filePath, buildTokenList(combinedTokens, version));
+}
+
+// Match a CoinMarket to a token from the Coingecko TokenList based on
+// matching symbol and Coingecko numeric ID (extracted from the image URL),
+// because symbols often collide.
+function findTokenFromList(
+  coinMarket: CoinMarket,
+  tokens: Token[]
+): Token | null {
+  const imageIdPattern = /\/coins\/images\/(\d+)\//;
+  const match = coinMarket.image.match(imageIdPattern);
+  if (!match) {
+    return null;
+  }
+  return tokens.find(
+    (token) =>
+      coinMarket.symbol.toLowerCase() === token.symbol.toLowerCase() &&
+      token.logoURI.match(imageIdPattern)?.[1] === match[1]
+  );
+}
+
+async function fetchTokens(coinMarkets: CoinMarket[], coingeckoAll: TokenList) {
+  const tokens: Token[] = [];
+
+  for (const coinMarket of coinMarkets) {
+    // The coin market data lacks some required fields (decimals, address, chain ID)
+    const { id, name } = coinMarket;
+
+    // If the coin market can be matched to a token in an existing token list,
+    // use that directly.
+    const tokenFromList = findTokenFromList(coinMarket, coingeckoAll.tokens);
+    if (tokenFromList) {
+      tokens.push(tokenFromList);
+      continue;
+    }
+
+    // Otherwise, fetch the coin data to get missing fields.
+    const coin = await client.coinId({
+      id,
+      sparkline: false,
+      tickers: false,
+      community_data: false,
+      market_data: false,
+      developer_data: false,
+      localization: false,
+    });
+
+    const {
+      // Fields such as name are also present in the coin data, but are
+      // unused in favour of the coin market data.
+      image: { thumb: logoURI },
+      detail_platforms,
+    } = coin as typeof coin & {
+      detail_platforms: {
+        [chain: string]: { decimal_place: number; contract_address: string };
+      };
+    };
+
+    // Try supported chains in order to get the contract address and decimals.
+    const chain = SUPPORTED_CHAINS.find(({ name }) => detail_platforms[name]);
+
+    if (!chain) {
+      // If there is no `detail_platforms` entry found, it is likely either not an
+      // ERC20, or not on a supported chain.
+      console.log(`Unsupported coin: ${id}`);
+      continue;
+    }
+
+    const { chainId } = chain;
+    const { decimal_place: decimals, contract_address: address } =
+      detail_platforms[chain.name];
+
+    if (decimals === 0 || address === "") {
+      // Some coins have a `detail_platforms` field but lack basic token data.
+      console.log(`Invalid token data: ${id}`);
```

Comment:
nit: can you add some context to the error so it's easier to see what was the issue? And maybe change from log to error

```suggestion
      console.warn(`Invalid token data: ${id}`, {  decimals, address });
```